src/pkcs11/mechanism.c

```diff
@@ -67,8 +67,7 @@ signature_data_buffer_append(struct operation_data *data,
 	if (in_len == 0)
 		return CKR_OK;
 
-	unsigned int new_len = data->buffer_len + in_len;
-	CK_BYTE *new_buffer = sc_mem_secure_alloc(new_len);
+	CK_BYTE *new_buffer = sc_mem_secure_alloc(data->buffer_len + in_len);
```

Comment:
Down the rabbit hole you can see a weird mix of length types that propagates to several places. Here, you have a mix of `unsigned int` and `CK_ULONG` (`unsigned long int`), `sc_mem_secure_alloc` takes `size_t`. Then you have `struct operation_data` whose `buffer_len` is `unsigned int`, but gets casted to `CK_ULONG` in `sc_pkcs11_signature_final`. Then there is `sc_pkcs11_verify_data` which uses `unsigned int`s, which themselves need a cast in `sc_pkcs11_verify_final`.

This looks like a big mess... Maybe we should stick to `CK_ULONG` on the pkcs#11 layer by
- changing `struct operation_data`'s `buffer_len` to `CK_ULONG` and
- changing `sc_pkcs11_verify_data`'s types to `CK_ULONG`

This would be my first suggestion, but I'm not exactly sure. In any case, I'd like to avoid fiddeling around at small pieces of the code while the big picture is not clear..

Reply:
Yeah. I think we can never avoid the casts. Even if we will sync on unsigned long in the pkcs11 layer, there will be cases where openssl or system calls will require or return `int`s, `size_t` or others and we will have to cast them back and forth. The point is that we should cast them in a sane way and copying the code back and forth will not solve this.

Any idea why these do not show up in either of the Linux builds? Is the MacOS build in master using some special build flags?

Reply:
I've updated the code to unify the pkcs#11 length/buffer types (a) and to explicitly check for an integer overflow (b). The second part may come handy for other pieces of the code as well.